nsqlookupd/lookup_protocol_v1.go

```diff
@@ -27,9 +28,30 @@ func (p *LookupProtocolV1) IOLoop(conn net.Conn) error {
 
 	client := NewClientV1(conn)
 	reader := bufio.NewReader(client)
+	readChan := make(chan string)
+	errChan := make(chan error)
+
 	for {
-		line, err = reader.ReadString('\n')
-		if err != nil {
+		exitLoop := false
+
+		// do this read in a goroutine so we can exit this loop if needed
+		go func() {
+			line, e := reader.ReadString('\n')
+			readChan <- line
+			errChan <- e
```

Comment:
I think it would be better to instead find a way to call `conn.Close()` when `exitChan` indicates. Here's one idea:

```go
doneChan := make(chan interface{})
defer close(doneChan)
go func() {
        select {
        case <-p.exitChan:
                conn.Close()
        case <-doneChan:
                break
        }
}
for {
        line, err = reader.ReadString('\n')
        ...
```

Here's a different way that does not involve an extra goroutine and channel for each connection:

```diff
--- a/nsqlookupd/nsqlookupd.go
+++ b/nsqlookupd/nsqlookupd.go
@@ -66,6 +66,7 @@ func (l *NSQLookupd) Main() error {
        tcpServer := &tcpServer{ctx: ctx}
        l.waitGroup.Wrap(func() {
                exitFunc(protocol.TCPServer(l.tcpListener, tcpServer, l.logf))
+               tcpServer.CloseAll()
        })
        httpServer := newHTTPServer(ctx)
        l.waitGroup.Wrap(func() {

--- a/nsqlookupd/tcp.go
+++ b/nsqlookupd/tcp.go
@@ -3,12 +3,14 @@ package nsqlookupd
 import (
        "io"
        "net"
+       "sync"
 
        "github.com/nsqio/nsq/internal/protocol"
 )
 
 type tcpServer struct {
        ctx *Context
+       conns sync.Map
 }
 
 func (p *tcpServer) Handle(clientConn net.Conn) {
@@ -41,9 +43,20 @@ func (p *tcpServer) Handle(clientConn net.Conn) {
                return
        }
 
+       p.conns.Store(clientConn.RemoteAddr(), clientConn)
+
        err = prot.IOLoop(clientConn)
        if err != nil {
                p.ctx.nsqlookupd.logf(LOG_ERROR, "client(%s) - %s", clientConn.RemoteAddr(), err)
-               return
        }
+
+       p.conns.Delete(clientConn.RemoteAddr())
+}
+
+func (p *tcpServer) CloseAll() {
+       p.conns.Range(func (k, v interface{}) bool {
+               v.(net.Conn).Close()
+               p.conns.Delete(k)
+               return true
+       })
 }
```

but I'm not sure if the overhead of the `sync.Map()` is something to worry about. probably not :)

Reply:
hmm, I think it would need a different way to call `tcpServer.CloseAll()` for my second proposal, ordering after `protocol.TcpServer()` returns seems like it shouldn't work

Reply:
a better way to call `CloseAll()` for that idea:

```diff
--- a/nsqlookupd/nsqlookupd.go
+++ b/nsqlookupd/nsqlookupd.go
@@ -18,6 +18,7 @@ type NSQLookupd struct {
        opts         *Options
        tcpListener  net.Listener
        httpListener net.Listener
+       tcpServer    *tcpServer
        waitGroup    util.WaitGroupWrapper
        DB           *RegistrationDB
 }
@@ -63,9 +64,9 @@ func (l *NSQLookupd) Main() error {
                })
        }
 
-       tcpServer := &tcpServer{ctx: ctx}
+       l.tcpServer = &tcpServer{ctx: ctx}
        l.waitGroup.Wrap(func() {
-               exitFunc(protocol.TCPServer(l.tcpListener, tcpServer, l.logf))
+               exitFunc(protocol.TCPServer(l.tcpListener, l.tcpServer, l.logf))
        })
        httpServer := newHTTPServer(ctx)
@@ -87,6 +88,7 @@ func (l *NSQLookupd) RealHTTPAddr() *net.TCPAddr {
 func (l *NSQLookupd) Exit() {
        if l.tcpListener != nil {
                l.tcpListener.Close()
+               l.tcpServer.CloseAll()
        }
```

Reply:
Trying that now, I like it. Thank you for taking a look at this!

Reply:
thanks for the updates! I'm going to take a day or two to test it a bit myself